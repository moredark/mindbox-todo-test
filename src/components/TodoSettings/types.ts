import { TodoSettings } from "../../types";

export interface TodoSettingsProps {
  settings: TodoSettings;
  onToggleAutoDelete: () => void;
}