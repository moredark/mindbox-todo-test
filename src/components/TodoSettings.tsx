import React, { useState, useRef } from 'react';
import { Box, Button, CheckBox, Drop, Text, Heading } from 'grommet';
import { Configure } from 'grommet-icons';
import { TodoSettings as TodoSettingsType } from '../types';

interface TodoSettingsProps {
  settings: TodoSettingsType;
  onToggleAutoDelete: () => void;
}

export const TodoSettings: React.FC<TodoSettingsProps> = ({ settings, onToggleAutoDelete }) => {
  const [showSettings, setShowSettings] = useState(false);
  const settingsButtonRef = useRef<HTMLButtonElement & HTMLAnchorElement>(null);

  return (
    <Box>
      <Button
        ref={settingsButtonRef}
        icon={<Configure color="text-strong" />}
        onClick={() => setShowSettings(!showSettings)}
        data-testid="settings-button"
      />

      {showSettings && settingsButtonRef.current && (
        <Drop
          target={settingsButtonRef.current}
          align={{ top: 'bottom', right: 'right' }}
          onClickOutside={() => setShowSettings(false)}
          onEsc={() => setShowSettings(false)}
          elevation="none"
        >
          <Box
            pad="medium"
            gap="small"
            width={{ min: '300px' }}
            data-testid="settings-panel"
            background="background-front"
            border={{ color: 'dark-3', size: '1px' }}
            round="small"
          >
            <Heading level={3} margin="none" size="small" color="text-strong">
              Settings
            </Heading>

            <Box margin={{ top: 'small' }}>
              <CheckBox
                label="Automatically delete completed tasks"
                checked={settings.autoDeleteCompleted}
                onChange={onToggleAutoDelete}
                data-testid="auto-delete-setting"
              />
              <Text
                size="small"
                color="dark-5"
                margin={{ top: 'xsmall' }}
              >
                When enabled, completed tasks will be automatically removed
              </Text>
            </Box>
          </Box>
        </Drop>
      )}
    </Box>
  );
};
