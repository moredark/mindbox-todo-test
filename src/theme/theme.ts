import { ThemeType } from 'grommet';
import { deepMerge } from 'grommet/utils';
import { dark } from 'grommet/themes';

export const accentColor = '#7D4CDB';

export const customTheme: ThemeType = deepMerge(dark, {
  global: {
    colors: {
      background: '#121212',
      'background-back': '#1E1E1E',
      'background-front': '#2D2D2D',
      'background-contrast': '#3D3D3D',
      text: '#EEEEEE',
      'text-strong': '#FFFFFF',
      'dark-6': '#999999',
      'dark-5': '#BBBBBB',
      'accent-1': accentColor,
      brand: accentColor
    },
    focus: {
      border: {
        color: accentColor
      }
    }
  },
  button: {
    border: {
      radius: '4px'
    },
    primary: {
      color: accentColor
    }
  },
  checkBox: {
    hover: {
      border: {
        color: accentColor
      }
    },
    color: accentColor
  }
});
