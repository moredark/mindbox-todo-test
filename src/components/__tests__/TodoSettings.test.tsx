import { render, screen, fireEvent } from '@testing-library/react';
import { Grommet } from 'grommet';
import { TodoSettings } from '../TodoSettings';
import { customTheme } from '../../theme/theme';

describe('TodoSettings', () => {
  const defaultProps = {
    settings: {
      autoDeleteCompleted: false
    },
    onToggleAutoDelete: jest.fn()
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  const renderComponent = (props = defaultProps) => {
    return render(
      <Grommet theme={customTheme} full>
        <TodoSettings {...props} />
      </Grommet>
    );
  };

  it('should match snapshot with settings closed', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with settings open', () => {
    renderComponent();
    const settingsButton = screen.getByTestId('settings-button');
    fireEvent.click(settingsButton);
    
    const settingsPanel = screen.getByTestId('settings-panel');
    expect(settingsPanel).toBeInTheDocument();
    expect(settingsPanel).toMatchSnapshot();
  });

  it('should toggle settings panel on button click', () => {
    renderComponent();
    const settingsButton = screen.getByTestId('settings-button');

    fireEvent.click(settingsButton);
    expect(screen.getByTestId('settings-panel')).toBeInTheDocument();

    fireEvent.click(settingsButton);
    expect(screen.queryByTestId('settings-panel')).not.toBeInTheDocument();
  });

  it('should call onToggleAutoDelete when checkbox is clicked', () => {
    renderComponent();
    const settingsButton = screen.getByTestId('settings-button');
    fireEvent.click(settingsButton);

    const checkbox = screen.getByTestId('auto-delete-setting');
    fireEvent.click(checkbox);

    expect(defaultProps.onToggleAutoDelete).toHaveBeenCalledTimes(1);
  });

  it('should show correct checkbox state', () => {
    renderComponent({
      settings: {
        autoDeleteCompleted: true
      },
      onToggleAutoDelete: defaultProps.onToggleAutoDelete
    });

    const settingsButton = screen.getByTestId('settings-button');
    fireEvent.click(settingsButton);

    const checkbox = screen.getByTestId('auto-delete-setting');
    expect(checkbox).toBeChecked();
  });
});
