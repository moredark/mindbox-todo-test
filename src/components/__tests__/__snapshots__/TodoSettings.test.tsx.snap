// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoSettings should match snapshot with settings closed 1`] = `
<div>
  <div
    class="StyledGrommet-sc-19lkkz7-0 huBofQ"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 gsnvgM"
    >
      <button
        class="StyledButton-sc-323bzc-0 cszWFM"
        data-testid="settings-button"
        type="button"
      >
        <svg
          aria-label="Configure"
          class="StyledIcon-sc-ofa7kd-0 iZPhJX"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 15c4.009-.065 7-3.033 7-7 0-3.012-.997-2.015-2-1-.991.98-3 3-3 3l-4-4s2.02-2.009 3-3c1.015-1.003 1.015-2-1-2-3.967 0-6.947 2.991-7 7 .042.976 0 3 0 3-1.885 1.897-4.34 4.353-6 6-2.932 2.944 1.056 6.932 4 4 1.65-1.662 4.113-4.125 6-6 0 0 2.024-.042 3 0z"
            fill="none"
            stroke="#000"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`TodoSettings should match snapshot with settings open 1`] = `
<div
  class="StyledBox-sc-13pk1d4-0 eRySLq"
  data-testid="settings-panel"
>
  <h3
    class="StyledHeading-sc-1rdh4aw-0 jqbxUT"
  >
    Settings
  </h3>
  <div
    class="StyledBox__StyledBoxGap-sc-13pk1d4-1 fWNJrO"
  />
  <div
    class="StyledBox-sc-13pk1d4-0 dwnQxZ"
  >
    <label
      class="StyledCheckBox__StyledCheckBoxContainer-sc-1dbk5ju-1 hUjAKd"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 gDkzsB StyledCheckBox-sc-1dbk5ju-6 eaiPwo"
      >
        <input
          class="StyledCheckBox__StyledCheckBoxInput-sc-1dbk5ju-2 kwqnhR"
          data-testid="auto-delete-setting"
          type="checkbox"
        />
        <div
          class="StyledBox-sc-13pk1d4-0 dmVmJK StyledCheckBox__StyledCheckBoxBox-sc-1dbk5ju-3 fQtXhU"
        />
      </div>
      <span>
        Automatically delete completed tasks
      </span>
    </label>
    <span
      class="StyledText-sc-1sadyjn-0 flkTRG"
    >
      When enabled, completed tasks will be automatically removed
    </span>
  </div>
</div>
`;
